# 01 Data Validation & Import

## Goal

Validate the structure of the weekly UK Top 100 chart dataset before importing it into the project database.

### Tasks

- Verify sheet names
- Verify chart format
- Verify column names
- Verify data types
- Import all weekly charts from 1983 onwards

In [1]:
from pathlib import Path

In [2]:
project_path = Path.cwd().parent
project_path

WindowsPath('c:/Users/sabur/Desktop/what-happened-to-music')

In [3]:
raw_path = project_path / "data" / "raw"
raw_path

WindowsPath('c:/Users/sabur/Desktop/what-happened-to-music/data/raw')

In [4]:
chart_file = raw_path / "top_100_songs_1952_to_2024.xlsx"

chart_file

WindowsPath('c:/Users/sabur/Desktop/what-happened-to-music/data/raw/top_100_songs_1952_to_2024.xlsx')

In [6]:
import pandas as pd
excel = pd.ExcelFile(chart_file)

In [7]:
excel.sheet_names

['Week 19521114',
 'Week 19521121',
 'Week 19521128',
 'Week 19521205',
 'Week 19521212',
 'Week 19521219',
 'Week 19521226',
 'Week 19530102',
 'Week 19530109',
 'Week 19530116',
 'Week 19530123',
 'Week 19530130',
 'Week 19530206',
 'Week 19530213',
 'Week 19530220',
 'Week 19530227',
 'Week 19530306',
 'Week 19530313',
 'Week 19530320',
 'Week 19530327',
 'Week 19530403',
 'Week 19530410',
 'Week 19530417',
 'Week 19530424',
 'Week 19530501',
 'Week 19530508',
 'Week 19530515',
 'Week 19530522',
 'Week 19530529',
 'Week 19530605',
 'Week 19530612',
 'Week 19530619',
 'Week 19530626',
 'Week 19530703',
 'Week 19530710',
 'Week 19530717',
 'Week 19530724',
 'Week 19530731',
 'Week 19530807',
 'Week 19530814',
 'Week 19530821',
 'Week 19530828',
 'Week 19530904',
 'Week 19530911',
 'Week 19530918',
 'Week 19530925',
 'Week 19531002',
 'Week 19531009',
 'Week 19531016',
 'Week 19531023',
 'Week 19531030',
 'Week 19531106',
 'Week 19531113',
 'Week 19531120',
 'Week 19531127',
 'Week 195

In [9]:
len(excel.sheet_names)

3725

In [10]:
excel.sheet_names[0]

'Week 19521114'

In [11]:
excel.sheet_names[-1]

'Week 20240329'

In [16]:
first_week = pd.read_excel(
    chart_file,
    sheet_name=excel.sheet_names[0]
)

first_week.head(15)

,Song,Artist,Position,Last Week,Peak,Weeks on Chart,Week
0,HE IN MY HEART,AL MARTINO,1,NaN,1,1,14 November 1952- 20 November 1952
1,YOU BELONG TO ME,JO STAFFORD,2,NaN,2,1,14 November 1952- 20 November 1952
2,SOMEWHE ALONG THE WAY,NAT 'KING' COLE,3,NaN,3,1,14 November 1952- 20 November 1952
3,THE ISLE OF INNISFE,BING CROSBY,4,NaN,4,1,14 November 1952- 20 November 1952
4,FEET UP (PAT HIM ON THE PO-PO),GUY MITCHELL,5,NaN,5,1,14 November 1952- 20 November 1952
5,HALF AS MUCH,ROSEMARY CLOONEY,6,NaN,6,1,14 November 1952- 20 November 1952
6,HIGH NOON (DO NOT FORSAKE ME),FRANKIE LAINE,7,NaN,7,1,14 November 1952- 20 November 1952
7,HIGH NOON (DO NOT FORSAKE ME),FRANKIE LAINE,7,NaN,7,1,14 November 1952- 20 November 1952
8,BLUE TANGO,RAY MARTIN,8,NaN,8,1,14 November 1952- 20 November 1952
9,BLUE TANGO,RAY MARTIN,8,NaN,8,1,14 November 1952- 20 November 1952


### First observations

- Weekly sheets have a consistent column structure.
- The first sheet contains duplicate rows for several chart positions.
- Needs further investigation before import.

In [15]:
first_week.info()

<class 'pandas.DataFrame'>
RangeIndex: 15 entries, 0 to 14
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Song            15 non-null     str    
 1   Artist          15 non-null     str    
 2   Position        15 non-null     int64  
 3   Last Week       0 non-null      float64
 4   Peak            15 non-null     int64  
 5   Weeks on Chart  15 non-null     int64  
 6   Week            15 non-null     str    
dtypes: float64(1), int64(3), str(3)
memory usage: 972.0 bytes


In [14]:
first_week.columns

Index(['Song', 'Artist', 'Position', 'Last Week', 'Peak', 'Weeks on Chart',
       'Week'],
      dtype='str')

In [17]:
week_1983 = pd.read_excel(
    chart_file,
    sheet_name="Week 19830107"
)

week_1983.head(15)

,Song,Artist,Position,Last Week,Peak,Weeks on Chart,Week
0,SAVE YOUR LOVE,RENEE AND RENATO,1,LW:1,1,11,2 January 1983- 8 January 1983
1,YOU CAN'T HURRY LOVE,PHIL COLLINS,2,LW:6,2,6,2 January 1983- 8 January 1983
2,A WINTER'S TALE,DAVID ESSEX,3,LW:7,3,5,2 January 1983- 8 January 1983
3,BEST YEARS OF OUR LIVES,MODERN ROMANCE,4,LW:8,4,9,2 January 1983- 8 January 1983
4,OUR HOUSE,MADNESS,5,LW:5,5,7,2 January 1983- 8 January 1983
5,TIME (CLOCK OF THE HEART),CULTURE CLUB,6,LW:4,3,7,2 January 1983- 8 January 1983
6,THE SHAKIN' STEVENS EP,SHAKIN' STEVENS,7,LW:2,2,5,2 January 1983- 8 January 1983
7,ORVILLE'S SONG,KEITH HARRIS AND ORVILLE,8,LW:33,8,4,2 January 1983- 8 January 1983
8,PEACE ON EARTH/LITTLE DRUMMER BOY,DAVID BOWIE AND BING CROSBY,9,LW:3,3,7,2 January 1983- 8 January 1983
9,ALL THE LOVE IN THE WORLD,DIONNE WARWICK,10,LW:14,10,5,2 January 1983- 8 January 1983


In [18]:
week_1983.info()

<class 'pandas.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   Song            100 non-null    str  
 1   Artist          100 non-null    str  
 2   Position        100 non-null    int64
 3   Last Week       100 non-null    str  
 4   Peak            100 non-null    int64
 5   Weeks on Chart  100 non-null    int64
 6   Week            100 non-null    str  
dtypes: int64(3), str(4)
memory usage: 5.6 KB


In [19]:
week_1983.columns

Index(['Song', 'Artist', 'Position', 'Last Week', 'Peak', 'Weeks on Chart',
       'Week'],
      dtype='str')

In [20]:
selected_sheets = [
    sheet
    for sheet in excel.sheet_names
    if sheet >= "Week 19830107"
]

len(selected_sheets)

2152

In [21]:
selected_sheets[:5]

['Week 19830107',
 'Week 19830114',
 'Week 19830121',
 'Week 19830128',
 'Week 19830204']

In [22]:
selected_sheets[-5:]

['Week 20240301',
 'Week 20240308',
 'Week 20240315',
 'Week 20240322',
 'Week 20240329']

In [24]:
test_charts = []

for sheet in selected_sheets[:5]:
    week_df = excel.parse(sheet)

    test_charts.append(week_df)

In [25]:
len(test_charts)

5

In [26]:
type(test_charts[0])

pandas.DataFrame

In [28]:
weekly_charts = []

for sheet in selected_sheets:
    week_df = excel.parse(sheet)
    weekly_charts.append(week_df)

In [29]:
len(weekly_charts)

2152

In [30]:
chart_history = pd.concat(
    weekly_charts,
    ignore_index=True
)

In [31]:
chart_history.shape

(211479, 7)

## Initial Data Validation

After importing all weekly UK chart sheets (1983–2024), several validation checks were performed:

- Verified the number of imported weeks.
- Checked the consistency of column names.
- Compared the number of chart entries per week.

In [32]:
chart_history.groupby("Week").size().describe()

count    2146.000000
mean       98.545666
std         7.799529
min        74.000000
25%       100.000000
50%       100.000000
75%       100.000000
max       200.000000
dtype: float64

In [33]:
chart_history.groupby("Week").size().value_counts().sort_index()

74        2
75      140
76        2
99       21
100    1975
150       2
200       4
Name: count, dtype: int64

In [34]:
songs_per_week = chart_history.groupby("Week").size()

songs_per_week[songs_per_week > 100]

Week
1 January 1989- 7 January 1989      200
2 January 2005- 8 January 2005      200
26 December 1993- 1 January 1994    150
26 December 1999- 1 January 2000    200
26 December 2010- 1 January 2011    200
3 January 1993- 9 January 1993      150
dtype: int64

### Data validation findings

- Most weeks contain exactly 100 chart entries.
- 140 early weeks contain 75 entries (expected historical format).
- Six weeks around the New Year contain duplicated chart positions.
- These weeks require special handling before final analysis.

In [38]:
week_1993 =chart_history[chart_history["Week"] == "26 December 1993- 1 January 1994"]
week_1993.head(15)

,Song,Artist,Position,Last Week,Peak,Weeks on Chart,Week
49851,MR. BLOBBY,MR. BLOBBY,1,LW:1,1,5,26 December 1993- 1 January 1994
49852,BABE,TAKE THAT,2,LW:2,1,3,26 December 1993- 1 January 1994
49853,TWIST AND SHOUT FT JACK RADICS AND TAXI GANG,CHAKA DEMUS AND PLIERS FEATURING JACK RADICS A...,3,LW:3,3,3,26 December 1993- 1 January 1994
49854,FOR WHOM THE BELL TOLLS,THE BEE GEES,4,LW:4,4,6,26 December 1993- 1 January 1994
49855,IT'S ALRIGHT,EAST 17,5,LW:5,5,5,26 December 1993- 1 January 1994
49856,THE PERFECT YEAR,DINA CARROLL,6,LW:7,6,4,26 December 1993- 1 January 1994
49857,I'D DO ANYTHING FOR LOVE (BUT I WON'T DO THAT),MEAT LOAF,7,LW:6,1,13,26 December 1993- 1 January 1994
49858,BAT OUT OF HELL {1993},MEAT LOAF,8,LW:8,8,10,26 December 1993- 1 January 1994
49859,TRUE LOVE,ELTON JOHN AND KIKI DEE,9,LW:9,2,7,26 December 1993- 1 January 1994
49860,COME BABY COME,K7,10,LW:13,10,4,26 December 1993- 1 January 1994


In [39]:
display(week_1993.shape)
week_1993["Position"].value_counts().sort_index()

(150, 7)

Position
1     2
2     2
3     2
4     2
5     2
     ..
71    2
72    2
73    2
74    2
75    2
Name: count, Length: 75, dtype: int64